src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+
+        /// <inheritdoc cref="StoreSettings"/>
+        private readonly StoreSettings storeSettings;
+
+        /// <inheritdoc cref="ConcurrentChain"/>
+        private readonly ConcurrentChain chain;
+
+        /// <inheritdoc cref="IBlockRepository"/>
+        private readonly IBlockRepository blockRepository;
+
+        /// <summary>Queue which contains blocks that should be saved to the database.</summary>
+        private readonly AsyncQueue<BlockPair> blocksQueue;
+
+        /// <summary>Task that runs <see cref="DequeueBlocksContinuouslyAsync"/>.</summary>
+        private Task dequeueLoopTask;
+
+        public BlockStore(
+            IBlockRepository blockRepository,
+            ConcurrentChain chain,
+            IChainState chainState,
+            StoreSettings storeSettings,
+            INodeLifetime nodeLifetime,
+            ILoggerFactory loggerFactory)
+        {
+            this.chainState = chainState;
+            this.nodeLifetime = nodeLifetime;
+            this.storeSettings = storeSettings;
+            this.chain = chain;
+            this.blockRepository = blockRepository;
+
+            this.blocksQueue = new AsyncQueue<BlockPair>();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <summary>
+        /// Initializes the <see cref="BlockStore"/>.
+        /// <para>
+        /// If StoreTip is <c>null</c>, the store is out of sync. This can happen when:</para>
+        /// <list>
+        ///     <item>The node crashed.</item>
+        ///     <item>The node was not closed down properly.</item>
+        /// </list>
+        /// <para>
+        /// To recover we walk back the chain until a common block header is found and set the <see cref="BlockStore"/>'s <see cref="StoreTip"/> to that.
+        /// </para>
+        /// </summary>
+        public async Task InitializeAsync()
+        {
+            this.logger.LogTrace("()");
+
+            if (this.storeSettings.ReIndex)
+                throw new NotImplementedException();
+
+            ChainedHeader initializationTip = this.chain.GetBlock(this.blockRepository.BlockHash);
+            this.SetStoreTip(initializationTip);
+            
+            if (this.StoreTip == null)
+                await this.RecoverStoreTipAsync().ConfigureAwait(false);
+
+            this.logger.LogDebug("Initialized block store tip at '{0}'.", this.StoreTip);
+
+            if (this.storeSettings.TxIndex != this.blockRepository.TxIndex)
+            {
+                if (this.StoreTip != this.chain.Genesis)
+                    throw new BlockStoreException("You need to rebuild the block store database using -reindex-chainstate to change -txindex");
+
+                if (this.storeSettings.TxIndex)
+                    await this.blockRepository.SetTxIndexAsync(this.storeSettings.TxIndex).ConfigureAwait(false);
+            }
+            
+            // Make sure that block store initializes before the consensus. 
```

Comment:
the code below doesn't actually `Make sure that block store initializes before the consensus` it just errors if it doesn't

Reply:
Fixed